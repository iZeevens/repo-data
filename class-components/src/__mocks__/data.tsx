import { ComicsData } from '../interfaces/searchTypes/searchTypes'
import { CardDetailsData } from '../interfaces/cardsTypes/cardsTypes';

const createComics = (
  title: string,
  publishedYear: number,
  publishedMonth: number,
  publishedDay: number,
  numberOfPages: number,
  photonovel: boolean,
  adaptation: boolean,
  uid: string
): ComicsData => ({
  title,
  publishedYear,
  publishedMonth,
  publishedDay,
  numberOfPages,
  photonovel,
  adaptation,
  uid,
})

const detailsCardsData = [
  {
    title: 'Card 1',
    numberOfPages: 100,
    stardateFrom: 500,
    stardateTo: 10000,
    yearFrom: 10,
    yearTo: 20,
  },
]

const cardDetails: CardDetailsData = {
  title: 'The Final Frontier',
  numberOfPages: 250,
  stardateFrom: 42000,
  stardateTo: 43000,
  yearFrom: 2365,
  yearTo: 2370,
  characters: [
    {
      name: 'James T. Kirk',
      gender: 'M',
      yearOfBirth: 2233,
      yearOfDeath: 2293,
    },
    {
      name: 'Spock',
      gender: 'M',
      yearOfBirth: 2230,
      yearOfDeath: 2293,
    },
    {
      name: 'Nyota Uhura',
      gender: 'F',
      yearOfBirth: 2239,
      yearOfDeath: 2293,
    },
  ],
  photonovel: true,
  adaptation: false,
  uid: 'CCMA0000076475',
};

const comicsData: ComicsData[] = [
  createComics('Card 1', 2024, 5, 15, 100, true, false, 'CCMA0000076475'),
  createComics('Card 2', 2024, 6, 20, 120, false, true, 'CCMA0000076476'),
  createComics('Card 3', 2024, 6, 20, 120, false, true, 'CCMA0000076476'),
  createComics('Card 4', 2024, 6, 20, 120, false, true, 'CCMA0000076476'),
  createComics('Card 5', 2024, 6, 20, 120, false, true, 'CCMA0000076476'),
  createComics('Card 6', 2024, 6, 20, 120, false, true, 'CCMA0000076476'),
]

export { detailsCardsData, cardDetails, comicsData }
