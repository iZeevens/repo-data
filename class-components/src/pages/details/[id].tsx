import { GetServerSideProps } from 'next'
import { CardDetailsData } from '../../interfaces/cardsTypes/cardsTypes'
import CloseBtn from '../../components/closeBtn/closeBtn'

interface DetailsPageProps {
  comicsData?: CardDetailsData | null
}

export default function DetailsPage({ comicsData }: DetailsPageProps) {
  if (!comicsData) {
    return <div>Data not found</div>
  }

  return (
    <div className="details-panel">
      <h3 className="details-title">{comicsData.title}</h3>
      <div className="details-info">
        <div className="details-info-about">
          <span>Number Pages: {comicsData.numberOfPages}</span>
          {comicsData.stardateFrom && (
            <span>Stardate From: {comicsData.stardateFrom}</span>
          )}
          {comicsData.stardateTo && (
            <span>Stardate To: {comicsData.stardateTo}</span>
          )}
          {comicsData.yearFrom && <span>Year From: {comicsData.yearFrom}</span>}
          {comicsData.yearTo && <span>Year To: {comicsData.yearTo}</span>}
          <span>Photonovel: {comicsData.photonovel ? 'Yes' : 'No'}</span>
          <span>Adaptation: {comicsData.adaptation ? 'Yes' : 'No'}</span>
        </div>
        {comicsData.characters && comicsData.characters.length > 0 ? (
          <div className="details-info--characters">
            <span>Characters:</span>
            {comicsData.characters.map((character, index) => (
              <div className="details-info--character" key={index}>
                <span>Name: {character.name}</span>
                <span>
                  Gender: {character.gender === 'M' ? 'Male' : 'Female'}
                </span>
                {character.yearOfBirth && (
                  <span>Year of birth: {character.yearOfBirth}</span>
                )}
                {character.yearOfDeath && (
                  <span>Year of death: {character.yearOfDeath}</span>
                )}
              </div>
            ))}
          </div>
        ) : (
          ''
        )}
      </div>
      <CloseBtn />
    </div>
  )
}

export const getServerSideProps: GetServerSideProps = async (context) => {
  const { id } = context.query
  if (typeof id !== 'string') {
    return { props: { comicsData: null } }
  }

  const res = await fetch(`https://stapi.co/api/v1/rest/comics?uid=${id}`)
  const comicsData = await res.json()

  return {
    props: {
      comicsData: comicsData.comics || null,
    },
  }
}
