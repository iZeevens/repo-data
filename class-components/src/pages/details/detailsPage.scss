.details-panel {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  max-width: 220px;
  border: 1px solid black;
  padding: 10px;

  .details-title {
    box-sizing: border-box;
    padding: 10px;
    margin: 0;
    width: 100%;
    border-bottom: 1px solid black;
  }

  .details-info {
    display: flex;
    flex-direction: column;

    .details-info-about {
      display: flex;
      flex-direction: column;
      gap: 3px;
      margin-top: 5px;
    }
  }

  .details-info--characters {
    display: flex;
    flex-direction: column;
    gap: 5px;
    margin-top: 20px;

    .details-info--character {
      display: flex;
      flex-direction: column;
      border: 1px solid black;
      padding: 5px;
    }
  }

  .details-panel-close {
    margin-top: 10px;
    width: 100px;
  }
}
