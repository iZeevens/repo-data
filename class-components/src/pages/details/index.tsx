// app/details/page.tsx

import './detailsPage.scss'
import { CardDetailsData } from '../../interfaces/cardsTypes/cardsTypes'
import CloseBtn from '../../components/closeBtn/closeBtn'

async function fetchComicsData(id: string): Promise<CardDetailsData | null> {
  const response = await fetch(`/api/comics/${id}`)
  console.log(id)
  if (response.ok) {
    return response.json()
  }
  return null
}

console.log('yes')

export default async function DetailsPage({
  params,
}: {
  params: { id: string }
}) {
  const { id } = params
  const comicsData = await fetchComicsData(id)

  if (!comicsData) {
    return <div>Data not found</div>
  }

  return (
    <div className="details-panel">
      <h3 className="details-title">{comicsData?.title}</h3>
      <div className="details-info">
        <div className="details-info-about">
          <span>Number Pages: {comicsData?.numberOfPages}</span>
          {comicsData?.stardateFrom && (
            <span>Stardate From: {comicsData?.stardateFrom}</span>
          )}
          {comicsData?.stardateTo && (
            <span>Stardate To: {comicsData.stardateTo}</span>
          )}
          {comicsData?.yearFrom && (
            <span>Year From: {comicsData.yearFrom}</span>
          )}
          {comicsData?.yearTo && <span>Year To: {comicsData.yearTo}</span>}
          <span>Photonovel: {comicsData?.photonovel ? 'Yes' : 'No'}</span>
          <span>Adaptation: {comicsData?.adaptation ? 'Yes' : 'No'}</span>
        </div>
        {comicsData?.characters && comicsData.characters.length > 0 ? (
          <div className="details-info--characters">
            <span>Characters:</span>
            {comicsData?.characters.map((character, index) => (
              <div className="details-info--character" key={index}>
                <span>Name: {character.name}</span>
                <span>
                  Gender: {character.gender === 'M' ? 'Male' : 'Female'}
                </span>
                {character.yearOfBirth && (
                  <span>Year of birth: {character.yearOfBirth}</span>
                )}
                {character.yearOfDeath && (
                  <span>Year of death:{character.yearOfDeath}</span>
                )}
              </div>
            ))}
          </div>
        ) : (
          ''
        )}
      </div>
      <CloseBtn />
    </div>
  )
}
