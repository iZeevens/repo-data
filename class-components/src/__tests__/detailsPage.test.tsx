import DetailsPage from '../pages/details/[id]'
import { render, screen } from '@testing-library/react'
import { ParsedUrlQuery } from 'querystring'
import { getServerSideProps } from '../pages/details/[id]'
import { GetServerSidePropsContext } from 'next'
import { cardDetails } from '../__mocks__/data'

vi.mock('next/navigation', () => ({
  useSearchParams() {
    return {
      get: vi.fn(),
    }
  },
  useRouter() {
    return {
      push: vi.fn(),
    }
  },
}))

describe('DetailsPage Testing', () => {
  it('check DOM', () => {
    render(DetailsPage({ comicsData: cardDetails }))
    expect(screen.getByText(/Number Pages:/)).toBeInTheDocument()
    expect(screen.getByText(/Characters:/)).toBeInTheDocument()
    expect(screen.getByText(/Nyota Uhura/)).toBeInTheDocument()
  })

  it('if data not exist', () => {
    render(DetailsPage({ comicsData: null }))
    expect(screen.getByText(/Data not found/)).toBeInTheDocument()
  })

  it('should return values if query parameters provided', async () => {
    const context = {
      query: { id: 'dsds' } as ParsedUrlQuery,
    }

    const value = await getServerSideProps(context as GetServerSidePropsContext)
    expect(value).toEqual({
      props: { comicsData: null },
    })
  })
})
