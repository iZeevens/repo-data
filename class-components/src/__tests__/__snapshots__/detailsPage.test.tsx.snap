// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DetailsPage Testing > params DetailsPage 1`] = `
<div>
  <div
    class="details-panel"
  >
    <h3
      class="details-title"
    >
      ... Like a Woman Scorned!
    </h3>
    <div
      class="details-info"
    >
      <div
        class="details-info-about"
      >
        <span>
          Number Pages: 
          22
        </span>
        <span>
          Stardate From: 
          6352.4
        </span>
        <span>
          Stardate To: 
          7935.6
        </span>
        <span>
          Photonovel: 
          No
        </span>
        <span>
          Adaptation: 
          No
        </span>
      </div>
      <div
        class="details-info--characters"
      >
        <span>
          Characters:
        </span>
        <div
          class="details-info--character"
        >
          <span>
            Name: 
            Nyota Uhura
          </span>
          <span>
            Gender: 
            Female
          </span>
          <span>
            Year of birth: 
            2239
          </span>
        </div>
        <div
          class="details-info--character"
        >
          <span>
            Name: 
            Hikaru Sulu
          </span>
          <span>
            Gender: 
            Male
          </span>
          <span>
            Year of birth: 
            2237
          </span>
        </div>
        <div
          class="details-info--character"
        >
          <span>
            Name: 
            Spock
          </span>
          <span>
            Gender: 
            Male
          </span>
          <span>
            Year of birth: 
            2230
          </span>
          <span>
            Year of death: 
            2285
          </span>
        </div>
        <div
          class="details-info--character"
        >
          <span>
            Name: 
            Janice Rand
          </span>
          <span>
            Gender: 
            Female
          </span>
          <span>
            Year of birth: 
            2242
          </span>
        </div>
        <div
          class="details-info--character"
        >
          <span>
            Name: 
            Montgomery Scott
          </span>
          <span>
            Gender: 
            Male
          </span>
          <span>
            Year of birth: 
            2222
          </span>
          <span>
            Year of death: 
            2267
          </span>
        </div>
        <div
          class="details-info--character"
        >
          <span>
            Name: 
            Pavel Chekov
          </span>
          <span>
            Gender: 
            Male
          </span>
          <span>
            Year of birth: 
            2245
          </span>
          <span>
            Year of death: 
            2401
          </span>
        </div>
        <div
          class="details-info--character"
        >
          <span>
            Name: 
            Christine Chapel
          </span>
          <span>
            Gender: 
            Female
          </span>
          <span>
            Year of birth: 
            2237
          </span>
          <span>
            Year of death: 
            2298
          </span>
        </div>
        <div
          class="details-info--character"
        >
          <span>
            Name: 
            James T. Kirk
          </span>
          <span>
            Gender: 
            Male
          </span>
          <span>
            Year of birth: 
            2233
          </span>
          <span>
            Year of death: 
            2293
          </span>
        </div>
        <div
          class="details-info--character"
        >
          <span>
            Name: 
            Leonard McCoy
          </span>
          <span>
            Gender: 
            Male
          </span>
          <span>
            Year of birth: 
            2227
          </span>
        </div>
        <div
          class="details-info--character"
        >
          <span>
            Name: 
            DiFalco
          </span>
          <span>
            Gender: 
            Female
          </span>
        </div>
      </div>
    </div>
    <button
      class="details-panel-close btn"
    >
      close
    </button>
  </div>
</div>
`;
